rust/noosphere-ns/tests/ns_test.rs

```diff
@@ -0,0 +1,192 @@
+#![cfg(not(target_arch = "wasm32"))]
+#![cfg(test)]
+pub mod utils;
+use anyhow::{anyhow, Result};
+use futures::future::try_join_all;
+use libipld_cbor::DagCborCodec;
+use noosphere_core::authority::generate_ed25519_key;
+use noosphere_ns::{
+    dht::DHTNode,
+    utils::{generate_capability, generate_fact},
+    NameSystem, NameSystemBuilder,
+};
+use noosphere_storage::{
+    db::SphereDb, encoding::derive_cid, memory::MemoryStorageProvider, memory::MemoryStore,
+};
+
+use ucan::{builder::UcanBuilder, crypto::KeyMaterial, time::now};
+use ucan_key_support::ed25519::Ed25519KeyMaterial;
+use utils::create_bootstrap_nodes;
+
+/// Data related to an owner sphere and a NameSystem running
+/// on its behalf in it's corresponding gateway.
+struct NSData {
+    pub ns: NameSystem<MemoryStore>,
+    pub owner_sphere_key: Ed25519KeyMaterial,
+    pub owner_sphere_id: String,
+}
+
+/// Generates a DHT network bootstrap node with `ns_count`
+/// NameSystems connected, each with a corresponding owner sphere.
+async fn generate_name_systems_network(
+    ns_count: usize,
+) -> Result<(DHTNode, SphereDb<MemoryStore>, Vec<NSData>)> {
+    let bootstrap_node = create_bootstrap_nodes(1)
+        .map_err(|e| anyhow!(e.to_string()))?
+        .pop()
+        .unwrap();
+    let bootstrap_addresses = vec![bootstrap_node.p2p_address().unwrap().to_owned()];
+
+    let store = SphereDb::new(&MemoryStorageProvider::default()).await?;
+    let mut name_systems: Vec<NSData> = vec![];
+
+    for _ in 0..ns_count {
+        let owner_sphere_key = generate_ed25519_key();
```

Comment:
What does this key represent? We have been using `owner` to refer to a key that is a user's key. A `sphere`'s key is typically "burned" right after it is created (and doesn't directly sign or verify anything).

Reply:
Good Q: This key is used only for the DHT node, which is needed for libp2p implementation, as it derives `PeerId` (therefore the node's `Multiaddr` address) and all the underlying crypto/validation between peers. This represents the Gateway key. Though currently there is no connection "gateway key" and "DHT key" semantically, unless we added e.g. gateway keys to prioritize DHT records (#124). It could be an ephemeral key generated by the Gateway, though we lose static addressing in the network (i.e. "known good nodes" would be less useful if a gateway used new keys on every run)

Reply:
Correction: The above refers to the `ns_key` -- the `owner_sphere_key` is the sphere the gateway is managing. With delegation working/testing, this could be updated to mimic that behavior